### PR TITLE
Rendering optimization(many files changed)

### PR DESCRIPTION
**Pull request by kkateq**
Main take aways after optimisation:
- there is a big update to the latest version, so run `yarn` to update local packages first 
- do not return empty arrays or objects as default values from selectors, return always `null` if there is nothing to return
- inherit all new components from ‘components/PureComponent` which performs deep state and props comparison

### CODE REVIEW

File: src/client/components/CodeDiffView/CodeDiffView.js [hunk 1]
```diff
@@ -21,22 +22,33 @@ export type Props = {
...
+
 class CodeDiffView extends PureComponent {
+  static defaultProps = {
+    viewType: '0',
+  }
+
   props: Props
 
+  shouldComponentUpdate(nextProps: Object, nextState: Object) {
+    return (this.props.rawDiff || this.props.unifiedDiff || this.props.sideBySideDiff) &&
```
Comment (wsprent) [inline src/client/components/CodeDiffView/CodeDiffView.js]: The component should update if
1. the current props contain at least of the kinds of diffs, _and_
2. the next props are different from the current.

Is 1. correct? Shouldn't it be if the _next_ props contain a diff?

File: src/client/components/List/index.js [hunk 2]
```diff
@@ -10,11 +11,15 @@ export type Props = {
   pageSize: number,
   total: number,
   activePage: number,
-  children?: number | string | React.Element <*> | Array < any >,
-  onPageSelect?: Function,
+  children?: number | string | React.Element<*> | Array < any >,
+  onPageSelect ?: Function,
```
Comment (wsprent) [inline src/client/components/List/index.js]: This sudden space looks weird to me especially considering  `children?` just above it.

File: src/client/components/PureComponent/index.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import React from 'react'
+import _ from 'lodash'
+
+class PureComponent extends React.PureComponent {
+  shouldComponentUpdate(nextProps: Object, nextState: Object) {
+    return !_.isEqual(this.props, nextProps) || !_.isEqual(this.state, nextState)
+  }
+}
+
+export const pureComponent = BaseComponent => class extends PureComponent {
```
Comment (wsprent) [inline src/client/components/PureComponent/index.js]: Is it really sound to wrap functional components in a class? I lived with the assumption that these were lighter somehow.

File: src/client/containers/PullRequestSummary/BuildSection/index.js [hunk 4]
```diff
@@ -1,34 +1,26 @@
...
   dispatch: Function,
   revision: string,
-  builds: Array<BuildType>,
 }
 
 export const getPRSourceStamp = (state: Object, props: Object): Object =>
   createSelector(
     getPullRequest,
-    (pr) => ({
+    pr => ({
```
Comment (wsprent) [inline src/client/containers/PullRequestSummary/BuildSection/index.js]: Just a small comment: I'm not sure our style rules are consistent with these parentheses. Sometimes single arguments have them, sometimes they don't.

File: src/client/components/List/index.js (hunk 2 above)
Comment (kkateq) [inline src/client/components/List/index.js]: Looks weird to me also ...(that's the drawback of automatic formatting)

File: src/client/components/CodeDiffView/CodeDiffView.js (hunk 1 above)
Comment (kkateq) [inline src/client/components/CodeDiffView/CodeDiffView.js]: yes, next props, that's mistake

File: src/client/components/List/index.js (hunk 2 above)
Comment (kkateq) [inline src/client/components/List/index.js]: so the problem in the line above `number | string | React.Element<*> | Array < any >`, probably eslint doesn't know how to recognize this flow construction

File: src/client/components/PureComponent/index.js (hunk 3 above)
Comment (kkateq) [inline src/client/components/PureComponent/index.js]: Yes. There is another example of this same wrapping in `component/SelectableList.js` but maybe written in more clear way

File: src/client/containers/PullRequestSummary/BuildSection/index.js (hunk 4 above)
Comment (kkateq) [inline src/client/containers/PullRequestSummary/BuildSection/index.js]: agree, but I think this is minor, the bigger thing is that eslint doesn't understand flow type fields if you pass object to a functional component as we do a lot with `props` ...

File: src/client/components/CodeDiffView/CodeDiffView.js (hunk 1 above)
Comment (kkateq) [inline src/client/components/CodeDiffView/CodeDiffView.js]: Fixed.

**Pull request merged**